### PR TITLE
Divdgmax slepc shift invert

### PR DESCRIPTION
**Pull request by Laar**
Background:

For DivDGMax we want to solve an eigenvalue problem of the form 
```Sx = ω^2Mx```
 for the lowest (positive) ω^2 and thus the smallest positive eigenvalues. Assuming that 0 is not an eigenvalue this is equivalent to solving 
```ω^(-2) Sx = Mx```
which then looks for the largest eigenvalues `ω^(-2)`. This latter approach is favourable from a numerical point of view, as the eigenvalues at the end of the spectrum converge first, which in the second case are exactly the wanted eigenvalues.

Change:
The old code directly implemented the second form. The update changes to using the first form, and letting the Spectral Transform (ST) functionality of Slepc internally transform it to using the second form. The advantage of this is threefold:
 1. The more interesting eigenvalues `ω^2` are returned
 2. The eigenvectors are normalized with respect to `M`, which is better than with respect to `S`.
 3. The eigenvalue problem is now Generalized Hermitian according, which could allow a more efficient solve and should ensure that the eigenvalues are purely real.

The PR is in draft till I can test whether 3. actually gives some benefit.

### CODE REVIEW

File: applications/DG-Max/Utils/Symmetry/Domain.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (JensWehner) [inline applications/DG-Max/Utils/Symmetry/Domain.h:1]: How does this fit in?

File: applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp [hunk 2]
```diff
@@ -50,6 +50,8 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 using namespace hpgem;
 
+// Helper function
+
```
Comment (JensWehner) [inline applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp:54]: ?

File: applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp [hunk 3]
```diff
@@ -185,15 +187,27 @@ class DivDGMaxEigenvalue<DIM>::SolverWorkspace {
...
-        // a spectral transformation.
-        // It may be possible to remove this relic with the current version of
-        // SLEPc.
-        err = EPSSetOperators(solver_, massMatrix_, stiffnessMatrix_);
+        err = EPSSetOperators(solver_, stiffnessMatrix_, massMatrix_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+
+        // The mass matrix is positive semidefinite, the stiffness matrix is
+        // Hermitian indefinite (k-phase-shifts are also Hermitian for real k).
+        err = EPSSetProblemType(solver_, EPS_GHEP);
```
Comment (JensWehner) [inline applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp:195]: do we have to bump the Slepc version in cmake?

**Review by JensWehner — COMMENTED** [review]
This makes it more clear, the tesselation argument class seems to not belong into this PR.

File: applications/DG-Max/Utils/Symmetry/Domain.h (hunk 1 above)
Comment (Laar) [reply applications/DG-Max/Utils/Symmetry/Domain.h:1]: Crap, that should not have been committed.

File: applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp (hunk 3 above)
Comment (Laar) [reply applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp:195]: No, all the needed functionality is long there in SLEPc. But probably nobody knew about it or something like that.

**Comment by Laar** [comment]
Fixed the issues. The GHEP does not seem to affect the performance much, so leaving that in.

**Review by JensWehner — APPROVED** [review]

**Pull request merged**